cmd/apiserver-watcher/azure_handler.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/golang/glog"
+	"github.com/vishvananda/netlink"
+)
+
+// Prevent hairpin traffic when the apiserver is up
+
+// As per the Azure documentation (https://docs.microsoft.com/en-us/azure/load-balancer/concepts//limitations),
+// if a backend is load-balanced to itself, then the traffic will be dropped.
+//
+// This is because the L3LB does DNAT, so while the outgoing packet has a destination
+// IP of the VIP, the incoming load-balanced packet has a destination IP of the
+// host. That means that it "sees" a syn with the source and destination
+// IPs of itself, and duly replies wit a syn-ack back to itself. However, the client
+// socket expects a syn-ack with a source IP of the VIP, so it drops the packet.
+//
+// The solution is to redirect traffic destined to the lb vip back to ourselves.
+
+type azureHandler struct {
+	vip       string
+	ip4tables *iptables.IPTables
+	ip6tables *iptables.IPTables
+	stopCh    chan struct{}
+}
+
+var _ handler = &azureHandler{}
+
+func newAzureHandler(vip string) (handler, error) {
+	ipt4, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv4),
+		iptables.Timeout(5),
+	)
+	if err != nil {
+		return nil, err
+	}
+	ipt6, err := iptables.New(
+		iptables.IPFamily(iptables.ProtocolIPv6),
+		iptables.Timeout(iptablesTimeout),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &azureHandler{
+		vip:       vip,
+		ip4tables: ipt4,
+		ip6tables: ipt6,
+		stopCh:    make(chan struct{}),
+	}, nil
+}
+
+// onFailure: either stop the routes service, or write downfile
+func (az *azureHandler) onFailure() error {
+	select {
+	case az.stopCh <- struct{}{}:
+		glog.V(4).Info("stopping iptables sync loop")
+	default:
+		glog.V(4).Info("no sync iptables running")
+	}
+	// wait for iptables sync loop to finish
+	time.Sleep(iptablesTimeout * time.Second)
+	return az.deleteRules()
+}
+
+// onSuccess: either start routes service, or remove down file
+func (az *azureHandler) onSuccess() error {
+	if err := az.syncRulesOnce(); err != nil {
+		return err
+	}
+	// bz1930457 delete stale conntrack entries to the apiserver
+	ip := net.ParseIP(az.vip)
+	if ip.IsGlobalUnicast() {
+		var family netlink.InetFamily
+		family = netlink.FAMILY_V4
+		if ip.To4() == nil {
+			family = netlink.FAMILY_V6
+		}
+		filter := &netlink.ConntrackFilter{}
+		filter.AddIP(netlink.ConntrackOrigDstIP, ip)
```

Comment:
I'm pretty sure this should be 'ConntrackReplySrcIP'

Reply:
> sh-4.4# cat /proc/net/nf_conntrack| grep SYN | grep 6443
> ipv4     2 tcp      6 118 SYN_SENT src=35.191.10.141 dst=10.0.0.2 sport=32768 dport=6443 [UNREPLIED] src=10.0.0.2 dst=35.191.10.141 sport=6443 dport=32768 mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
> ipv4     2 tcp      6 115 SYN_SENT src=35.191.10.143 dst=10.0.0.2 sport=32768 dport=6443 [UNREPLIED] src=10.0.0.2 dst=35.191.10.143 sport=6443 dport=32768 mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2
> ipv4     2 tcp      6 119 SYN_SENT src=35.191.10.135 dst=10.0.0.2 sport=32768 dport=6443 [UNREPLIED] src=10.0.0.2 dst=35.191.10.135 sport=6443 dport=32768 mark=0 secctx=system_u:object_r:unlabeled_t:s0 zone=0 use=2

this runs before the redirect are working, so it flush everything targeting the VIPs, but maybe `ConntrackReplySrcIP` is safer indeed

Reply:
changed, the name is different in the netlink library
```	
ConntrackNatSrcIP         // -src-nat ip    Source NAT ip
```